c++/error-handling.md

```diff
@@ -0,0 +1,337 @@
+## Dynamically-Enforceable Contracts
+
+Dynamically-Enforceable API Contracts and the ISO C++ Standard
+form a group of contracts whose violation represents run-time bugs.
+Of these, a subset can be expressed formally within the program
+using assertions (including API preconditions and postconditions).
+
+These expressions can be evaluated at run-time but typically incur a cost
+which violates the zero-overhead principle in the case of a bug-free program.
+
+Broadly, there are three possible strategies that can be adopted
+at the point where an assertion would evaluate to `false`:
+
+### Log-And-Continue Enforcement Strategy
+
+The assert statement could emit a run-time diagnostic regarding the contract violation
+and then continue past the assertion.
+
+This is unwise: the program is known to exhibit undefined behaviour at this
+point. If the provider were able to identify a reason why a violation might have
+occurred they would have identified a bug. If the provider had identified a bug,
+they should have fixed the bug. If the provider had fixed the bug,
+they would no longer be able to identify a reason why a violation might have occurred.
+Thus any violation is beyond the comprehension of the provider and "all bets are
+off" regarding the behaviour of the program. This is essentially what UB means.
+
+Nevertheless, under certain constraints
+
+### Trap Enforcement Strategy
+
+The assert statement can halt the program.
+
+This has the disadvantage that the program completely fails to perform its task.
+But the advantage is that no further incorrect behaviour
+will be exhibited by the program.
+
+### Prevention Enforcement Strategy
+
+Assuming bugs are eliminated, no assert expression will ever be false.
+This means that assert statements behave as if they are not there.
+This in turn means that they can be empty or used as hints to the optimiser.
+
+There is concern ([P2064](https://wg21.link/p2064r0)) that assuming no violations
+is somehow incorrect. This is founded on the acceptance of bugs in production code
+and the assumption that optimisers *cause* undefined behaviour. They do not. Bugs
+cause undefined behaviour. If developers wish to release code with bugs to production,
+they are welcome to disable such assumptions. But they are invited, instead, to follow
+practices described as part of the Test User Contract which should help identify
+such bugs early on.
+
+***The author strongly encourages the use of assumptions in the implementation of
+assertions in order to unify the set of Dynamically-Enforceable Contracts and to
+simplify the implementation of the Test User Contract.***
+In practice, this entails marking violated assertions as unreachable
+
+```c++
+#define ASSERT(cond) ((cond) ? static_cast<void>(0) : __builtin_unreachable())
+```
+
+which is semantically sound, given violations already must never occur in correctly-written
+C++ programs.
+
+This strategy provides the maximum leeway to support enforcement profiles (below)
```

Comment:
I do not think that implementing contracts annotations with `__builtin_unreachable()` is necessary for enabling profilers.
The provision that upon failed runtime check a violation handler is invoked is enough to implement a feature in the UB-sanitizer that detects not only Abstract Machine Violations but also user contrct violations.

Constant expressions can also be handled directly: we could say that if the evaluation of a core constant expression invokes the violation handler, the program is ill formed.

So the only benefit that `__builtin_unreachable()` would give is the potential to implement tricky optimizations.
Personally, I am not against such optimizations, but I do not see a need to actively support them in the language. If such optimizations are really practical, compilers will implement them anyway, even if this is not standard-compliant. But I think it is standard-compliant if you install a violation handler with `__builtin_unreachable()`.

Reply:
> I do not think that implementing contracts annotations with `__builtin_unreachable()` is necessary for enabling profilers.

Do you mean a program profiling tool? I'm referring to *Enforcement Profiles* as listed below. These are different common use cases with which I hope to cover most users' needs. Perhaps I should choose a different term for these.

> The provision that upon failed runtime check a violation handler is invoked is enough to implement a feature in the UB-sanitizer that detects not only Abstract Machine Violations but also user contrct violations.

A violation handler isn't "enough", but excessive. C++20 using an `ASSERT` macro already does everything you describe. 

Are you saying that a handler (say, the one in P2339R0) would be invoked on violation of an ISO C++ Standard contract violation, e.g. divide-by-zero or null terminator dereference? I wasn't aware of such a suggestion. A C++20 sanitizer will trap such a violation. I wasn't aware that a P2339R0 violation handler would.

IOW, future proposals are a step backwards in terms of usability: the landscape will be bifurcated between handler-based violation and UB sanitizers. Inserting `__builtin_unreachable()` into a handler then becomes extra effort the user needs to go to -- just to get back to where she was in C++20.

> Constant expressions can also be handled directly: we could say that if the evaluation of a core constant expression invokes the violation handler, the program is ill formed.

Again, the C++20 user already enjoys compile-time diagnostics. Handler proposals offers the user something they already have -- at additional effort.

> So the only benefit that `__builtin_unreachable()` would give is the potential to implement tricky optimizations.
Personally, I am not against such optimizations, but I do not see a need to actively support them in the language. If such optimizations are really practical, compilers will implement them anyway, even if this is not standard-compliant. 

We should not be explicitly trying to optimise, and `__builtin_unreachable()` does not do that. But the toolchain user should continue to agree that "-O1 means that bugs are fixed, including all Dynamically-Enforceable Contracts" and that Dynamically-Enforceable API Contracts are a new category of these.

> But I think it is standard-compliant if you install a violation handler with `__builtin_unreachable()`.

I'm not opposed to such a handler, but I don't think that they add anything to the status quo. I see them as something needed to satisfy users who wish to log [BizBugs](http://open-std.org/JTC1/SC22/WG21/docs/papers/2021/p2358r0.pdf#7) without subjecting them to the rigour of a *true* contract violation.

---

I'll try to incorporate some of this into the discussion section somehow, which is the least baked section so far. I've got to think about the points I'm trying to make and the audience I'm trying to reach. Thanks for looking at this!

Reply:
My apologies. I think I typed this too late in the night, and I introduced an unnecessary confusion. You wrote, "Enforcement Profiles", I read it as "UB-sanitizers" and wrote back "profilers". The remainder of my comment had better be ignored.

Reply:
Let me take another attempt at communicating my thoughts.

First. It is not clear why it is called "Prevention Enforcement Strategy", how is using `__builtin_unreachable()` more preventive than the other listed strategies?

Second. I think that [P2064](https://wg21.link/p2064r0) doesn't do a good job at communicating legitimate concerns about turning contract violations into Abstract Machine Violations. I will try to present my arguments in a subsequent comment.

Third. I read this article as saying that the semantics of `__builtin_unreachable()` gives implementations enough power to implement all the Enforcement Profiles that you list. (I agree.) Now, in the defense of the current proposals, such as [P2182](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2182r1.html), they use a more complex specification in order to (1) allow the same (and more) Enforcement Profiles but (2) make sure that optimizations based on contract annotations is not one of them. They use terms like "install a violation handler (in an implementation-defined way)", but this does not mean that someone will have to define a function with a concrete signature.  For instance, a GCC could give an option to compile with UB-sanitizer, and this UB-sanitizer reports both Abstact Machine Violations and business contract violations (expressed with contract annotations). This happens with one compiler switch `-fsanitize=undefined`. Such compiler feature would be compliant with [P2182](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2182r1.html): the compiler switch would be this "implementation-defined way" of installing the handler. IOW, the burden is on the people who read the standardese, not on the programmers who use the compiler interface.

Reply:
> First. It is not clear why it is called "Prevention Enforcement Strategy", how is using __builtin_unreachable() more preventive than the other listed strategies?

Anything that aids testing is preventative and trapping is the best approach here so `__builtin_unreachable()` in combination with a sanitizer is as good as anything else. OTOH, log-and-continue is likely to be less preventative. It's like the difference between errors and warnings: logging messages can be ignored.

> IOW, the burden is on the people who read the standardese, not on the programmers who use the compiler interface.

It's not 100% clear to me. Does P2064 propose wording to support custom handlers in a portable way or does it simply leave the door open for implementation-defined handlers?

Reply:
> Anything that aids testing is preventative

Agreed so far.

> and trapping is the best approach here

This is confusing. My comment is regarding section "Prevention Enforcement Strategy". You argue that trapping is the best enforcement strategy. But there is a previous section "Trap Enforcement Strategy". Your comment explains the previous section well, but does not explain the current one ("Enforcement"). 

> so `__builtin_unreachable()` in combination with a sanitizer is as good as anything else.

If you used `__builtin_trap()` for assisting the sanitizer, I would be on the same page. But you use `__builtin_unreachable()`, which I believe does not have the trapping semantics. My experience (with one of the UB-sanitizers, and I think clang static analyzer as well) is that even these tools take the things you declare with `__builtin_unreachable()` for granted and do not try to sanitize/detect if what you assert with it is incorrect.

> OTOH, log-and-continue is likely to be less preventative.

I agree. But I thing this is beside the point of disagreement. My point of confusion is: what do you consider to be the difference between "Trap Enforcement Strategy" and "Prevention Enforcement Strategy"? Why is `__builtin_unreachable()` more preventive than `__builtin_trap()`?


> > IOW, the burden is on the people who read the standardese, not on the programmers who use the compiler interface.

> It's not 100% clear to me. Does P2064 propose wording to support custom handlers in a portable way or does it simply leave the door open for implementation-defined handlers?

P2064 is "Assumptions". The recent papers that propose solutions are

* [P2182R1](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2182r1.html) "Contract Support: Defining the Minimum Viable Feature Set",
* [P2388R0](https://isocpp.org/files/papers/P2388R0.html) "Abort-only contract support" (submitted only Today).

They both allow (but not require) the implementations to provide a way to install a custom violation handler. How the interface for installing such a handler looks like is not defined: it is left for implementations to decide. I can say it with confidence, because I co-authored both. But it is possible that this intent is not reflected clearly enough in these papers.

(The difference between the two papers is that the later does not allow throwing or continuing violation handlers.)

Reply:
Too many thread!

> If you used __builtin_trap() for assisting the sanitizer, I would be on the same page. But you use __builtin_unreachable(), which I believe does not have the trapping semantics. My experience (with one of the UB-sanitizers, and I think clang static analyzer as well) is that even these tools take the things you declare with __builtin_unreachable() for granted and do not try to sanitize/detect if what you assert with it is incorrect.

The reason Test User Contract is so difficult to convey is because it changes the rules. Trapping is desired in one case (Test User Contract) but not necessarily in another (Release, End User Contract). Assuming sanitizers are enabled in test builds, `__builtin_unreachable()` does a different thing in both cases but it also does the *right* thing in both cases.

I've rewritten the Test Case Contract section. I hope it's closer to helpful.

> what do you consider to be the difference between "Trap Enforcement Strategy" and "Prevention Enforcement Strategy"?

"Trap Enforcement Strategy":
* In a production build:
  * program **traps on** C++ API Contract violation
  * program assumes no ISO C++ Standard violation
* In a test build (w. all the sanitizer goodness):
  * program traps on C++ API Contract violation
  * program traps on ISO C++ Standard violation

Prevention Enforcement Strategy:
* In a production build:
  * program **assumes no** C++ API Contract violation
  * program assumes no ISO C++ Standard violation
* In a test build (w. all the sanitizer goodness):
  * program traps on C++ API Contract violation
  * program traps on ISO C++ Standard violation

A program author already **must** sanitize before they optimise because the program author needs to prevent UB. That's a C++20 fact of life. With the Prevention Enforcement Strategy, the same rule applies to C++ API Contracts. They must be ruled out.

> Why is __builtin_unreachable() more preventive than __builtin_trap()?

If you aim to prevent bugs before production, it's moot that bugs trap because no bugs -> no trapping. Therefore `__builtin_unreachable()` is the right choice because it goes away in a non-sanitizer build and produces more efficient code.

But if you didn't fix your bugs before production, you really need to keep the bug-handling code there somehow. (Depends on the Enforcement Profile.)

> I can say it with confidence, because I co-authored both. But it is possible that this intent is not reflected clearly enough in these papers.

I am largely on-board with those papers but I just don't think my view of software development is familiar beyond my ~victims~ colleagues. ;P I can believe it's alien to Bloomberg devs, given John's and Joshua's very extensive set of macros and Microsoft devs, given how many bugs the developers in p2064r0 let into production. I'm not saying my code is anywhere near perfect but I find it hard to believe those devs used sanitizers. I don't think that's viable any more.